File: app/models/location.py
from __future__ import annotations
from sqlalchemy import Boolean, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base

class LocationState(Base):
    __tablename__ = "location_states"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    teryt_code: Mapped[str | None] = mapped_column(String(16), nullable=True, unique=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    districts: Mapped[list["LocationDistrict"]] = relationship(back_populates="state")

class LocationDistrict(Base):
    __tablename__ = "location_districts"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    state_id: Mapped[int] = mapped_column(ForeignKey("location_states.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    teryt_code: Mapped[str | None] = mapped_column(String(16), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    state: Mapped["LocationState"] = relationship(back_populates="districts")
    cities: Mapped[list["LocationCity"]] = relationship(back_populates="district")

class LocationCity(Base):
    __tablename__ = "location_cities"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    district_id: Mapped[int] = mapped_column(ForeignKey("location_districts.id"), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    teryt_code: Mapped[str | None] = mapped_column(String(16), nullable=True)
    commune_code: Mapped[str | None] = mapped_column(String(8), nullable=True)
    commune_type: Mapped[str | None] = mapped_column(String(4), nullable=True)
    is_managed: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    district: Mapped["LocationDistrict"] = relationship(back_populates="cities")
    streets: Mapped[list["LocationStreet"]] = relationship(back_populates="city", cascade="all, delete-orphan")

class LocationStreet(Base):
    __tablename__ = "location_streets"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    city_id: Mapped[int] = mapped_column(ForeignKey("location_cities.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    teryt_code: Mapped[str | None] = mapped_column(String(16), nullable=True)
    city: Mapped["LocationCity"] = relationship(back_populates="streets")

File: crm-portal-ts/src/models/location.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, OneToMany, JoinColumn } from "typeorm";

@Entity("location_states")
export class LocationState {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", name: "teryt_code", length: 16, nullable: true, unique: true })
    terytCode?: string;

    @Column({ type: "boolean", name: "is_active", default: false })
    isActive!: boolean;

    @OneToMany(() => LocationDistrict, (district) => district.state)
    districts!: LocationDistrict[];
}

@Entity("location_districts")
export class LocationDistrict {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "state_id" })
    stateId!: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", name: "teryt_code", length: 16, nullable: true })
    terytCode?: string;

    @Column({ type: "boolean", name: "is_active", default: false })
    isActive!: boolean;

    @ManyToOne(() => LocationState, (state) => state.districts, { onDelete: "CASCADE" })
    @JoinColumn({ name: "state_id" })
    state!: LocationState;

    @OneToMany(() => LocationCity, (city) => city.district)
    cities!: LocationCity[];

    @OneToMany(() => LocationCommune, (commune) => commune.district)
    communes!: LocationCommune[];
}

@Entity("location_communes")
export class LocationCommune {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "district_id" })
    districtId!: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", name: "teryt_code", length: 16, nullable: true })
    terytCode?: string;

    @Column({ type: "varchar", name: "commune_code", length: 8, nullable: true })
    communeCode?: string;

    @Column({ type: "varchar", name: "commune_type", length: 4, nullable: true })
    communeType?: string;

    @Column({ type: "boolean", name: "is_managed", default: false })
    isManaged!: boolean;

    @Column({ type: "boolean", name: "is_default", default: false })
    isDefault!: boolean;

    @Column({ type: "boolean", name: "is_active", default: false })
    isActive!: boolean;

    @ManyToOne(() => LocationDistrict, (district) => district.communes, { onDelete: "CASCADE" })
    @JoinColumn({ name: "district_id" })
    district!: LocationDistrict;

    @OneToMany(() => LocationCity, (city) => city.commune)
    cities!: LocationCity[];

    @OneToMany(() => LocationStreet, (street) => street.commune)
    streets!: LocationStreet[];
}

@Entity("location_cities")
export class LocationCity {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "district_id" })
    districtId!: number;

    @Column({ type: "integer", name: "commune_id", nullable: true })
    communeId?: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", name: "teryt_code", length: 16, nullable: true })
    terytCode?: string;

    @Column({ type: "varchar", name: "commune_code", length: 8, nullable: true })
    communeCode?: string;

    @Column({ type: "varchar", name: "commune_type", length: 4, nullable: true })
    communeType?: string;

    @Column({ type: "boolean", name: "is_managed", default: false })
    isManaged!: boolean;

    @Column({ type: "boolean", name: "is_default", default: false })
    isDefault!: boolean;

    @Column({ type: "boolean", name: "is_active", default: false })
    isActive!: boolean;

    @ManyToOne(() => LocationDistrict, (district) => district.cities, { onDelete: "CASCADE" })
    @JoinColumn({ name: "district_id" })
    district!: LocationDistrict;

    @ManyToOne(() => LocationCommune, (commune) => commune.cities, { onDelete: "SET NULL", nullable: true })
    @JoinColumn({ name: "commune_id" })
    commune?: LocationCommune | null;

    @OneToMany(() => LocationStreet, (street) => street.city)
    streets!: LocationStreet[];
}

@Entity("location_streets")
export class LocationStreet {
    @PrimaryGeneratedColumn()
    id!: number;

    @Column({ type: "integer", name: "city_id" })
    cityId!: number;

    @Column({ type: "integer", name: "commune_id", nullable: true })
    communeId?: number;

    @Column({ type: "varchar", length: 128 })
    name!: string;

    @Column({ type: "varchar", name: "teryt_code", length: 16, nullable: true })
    terytCode?: string;

    @ManyToOne(() => LocationCity, (city) => city.streets, { onDelete: "CASCADE" })
    @JoinColumn({ name: "city_id" })
    city!: LocationCity;

    @ManyToOne(() => LocationCommune, (commune) => commune.streets, { onDelete: "SET NULL", nullable: true })
    @JoinColumn({ name: "commune_id" })
    commune?: LocationCommune | null;
}
